src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -1,79 +1,244 @@
+class GoogleDataprocDAO(protected val dataprocConfig: DataprocConfig, protected val proxyConfig: ProxyConfig)(implicit val system: ActorSystem, val executionContext: ExecutionContext)
   extends DataprocDAO with GoogleUtilities {
 
   private val httpTransport = GoogleNetHttpTransport.newTrustedTransport
   private val jsonFactory = JacksonFactory.getDefaultInstance
-  private val cloudPlatformScopes = List(PubsubScopes.CLOUD_PLATFORM)
+  private lazy val cloudPlatformScopes = List(ComputeScopes.CLOUD_PLATFORM)
+  private lazy val storageScopes = List(StorageScopes.DEVSTORAGE_FULL_CONTROL, ComputeScopes.COMPUTE, PlusScopes.USERINFO_EMAIL, PlusScopes.USERINFO_PROFILE)
+  private lazy val vmScopes = List(ComputeScopes.COMPUTE, ComputeScopes.CLOUD_PLATFORM)
+  private lazy val serviceAccountPemFile = new File(dataprocConfig.configFolderPath, dataprocConfig.serviceAccountPemName)
 
   private lazy val dataproc = {
-    new Dataproc.Builder(GoogleNetHttpTransport.newTrustedTransport,
-      JacksonFactory.getDefaultInstance, getDataProcServiceAccountCredential)
+    new Dataproc.Builder(httpTransport, jsonFactory, getServiceAccountCredential(cloudPlatformScopes))
       .setApplicationName("dataproc").build()
   }
-  private def getDataProcServiceAccountCredential: Credential = {
+
+  private lazy val storage = {
+    new Storage.Builder(httpTransport, jsonFactory, getServiceAccountCredential(storageScopes))
+      .setApplicationName("storage").build()
+  }
+
+
+  private def getServiceAccountCredential(scopes: List[String]): Credential = {
     new GoogleCredential.Builder()
       .setTransport(httpTransport)
       .setJsonFactory(jsonFactory)
       .setServiceAccountId(dataprocConfig.serviceAccount)
-      .setServiceAccountScopes(cloudPlatformScopes.asJava)
-      .setServiceAccountPrivateKeyFromPemFile(new java.io.File(dataprocConfig.serviceAccountPemPath))
+      .setServiceAccountScopes(scopes.asJava)
+      .setServiceAccountPrivateKeyFromPemFile(serviceAccountPemFile)
       .build()
   }
 
+  /* Creates a cluster in the given google project:
+       - Add a firewall rule to the user's google project if it doesn't exist, so we can access the cluster
+       - Create the initialization bucket for the cluster in the leo google project
+       - Upload all the necessary initialization files to the bucket
+       - Create the cluster in the google project
+     Currently, the bucketPath of the clusterRequest is not used - it will be used later as a place to store notebook results */
+  def createCluster(googleProject: GoogleProject, clusterName: String, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[ClusterResponse] = {
+    // Create the firewall rule in the google project if it doesn't already exist, so we can access the cluster
+    updateFirewallRules(googleProject)
```

Comment:
I think we're mixing synchronous and asynchronous calls here. Currently, `updateFirewallRules` does the firewall get synchronously, the firewall add asynchronously, and the firewall add Future is not waited upon before creating the bucket. I think we should do everything in terms of Future/flatMap. Something like this:

```
 def createCluster(googleProject: GoogleProject, clusterName: String, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[ClusterResponse] = {
    val bucketName = s"${clusterName}-${UUID.randomUUID.toString}"

    for {
      _ <- updateFirewallRules(googleProject)
      _ <- initializeBucket(dataprocConfig.leoGoogleBucket, clusterName, bucketName)
      operation <- build(googleProject, clusterName, clusterRequest, bucketName)
    } yield {
      val metadata = operation.getMetadata
      ClusterResponse(clusterName, googleProject, metadata.get("clusterUuid").toString, metadata.get("status").toString, metadata.get("description").toString, operation.getName)
    }
  }

  private def updateFirewallRules(googleProject: String): Future[Unit] = {
    Future {
      val request = new Compute(httpTransport, jsonFactory, getServiceAccountCredential(vmScopes)).firewalls().get(googleProject, dataprocConfig.clusterFirewallRuleName)
      executeGoogleRequest(request)
    }.recoverWith {
      case e: GoogleJsonResponseException if e.getStatusCode == 404 =>
        addFirewallRule(googleProject)
      case _ =>
        throw new FirewallRuleInaccessibleException(googleProject, dataprocConfig.clusterFirewallRuleName)
    }.mapTo[Unit]
  }
```